memory: realloc: defer to malloc if ptr is null

bhalevy opened the pull request:
As required by https://en.cppreference.com/w/cpp/memory/c/realloc:
> If ptr is a null pointer, the behavior is the same as calling std::malloc(new_size).

Fixes #978

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

raphaelsc commented on src/core/memory.cc:
```diff
@@ -1890,7 +1890,9 @@ void* realloc(void* ptr, size_t size) {
     if (try_trigger_error_injector()) {
         return nullptr;
     }
-    if (ptr != nullptr && !is_seastar_memory(ptr)) {
```
"If ptr is a null pointer, the behavior is the same as calling std::malloc(new_size)."

sounds like a good comment to place here, so future readers can understand why we're bypassing the realloc impl when ptr == null.

bhalevy replied:
Fixed in v2

raphaelsc commented on src/core/memory.cc:
```diff
@@ -1890,7 +1890,9 @@ void* realloc(void* ptr, size_t size) {
-    if (ptr != nullptr && !is_seastar_memory(ptr)) {
+    if (ptr == nullptr) {
+        return malloc(size);
+    } else if (!is_seastar_memory(ptr)) {
```
below there's the following comment:

"// if we're here, it's either ptr is a seastar memory ptr
    // or a nullptr, or, original functions aren't available"

we've to update it to "if we're here, it's a non-null seastar memory ptr..."

bhalevy replied:
Fixed in v2

raphaelsc reviewed (changes requested):
looks good to me, just had a minor comment on comment becoming outdated.

bhalevy commented:
In v2 (b5953d05):
- added comment for deferring to malloc
- corrected comment "if we're here" according to https://github.com/scylladb/seastar/pull/1100#discussion_r900540217, thanks.

raphaelsc reviewed (approved):
LGTM

nyh commented on tests/unit/alloc_test.cc:
```diff
@@ -175,4 +175,23 @@ SEASTAR_TEST_CASE(test_foreign_function_use_glibc_malloc) {
+SEASTAR_TEST_CASE(test_realloc_nullptr) {
+    auto p0 = realloc(test_nullptr, 8);
+    BOOST_REQUIRE(p0 != nullptr);
+    BOOST_REQUIRE_EQUAL(realloc(p0, 0), nullptr);
```
Nice. This is an important point that's easily missed - realloc(p0, 0) - when p0!=0, should do free(p). So good you tested it as well.
